File: liuy363/h10/h10.clj
(load-file "collection.clj") 

(defn summingPairs [xs sum]
(print("dfgdfda"))
  (letfn [(summingPairsHelper [xs the_pairs]
            (if (empty? xs) the_pairs
            
                (let [[fst & rest] xs]
                
                   (recur
                   rest
                   (doall 
                    (concat the_pairs
                            (for [snd rest 
                                  :when (<= (+ fst snd) sum)]
                              [fst snd]))))))
           (let [first_half  (future (summingPairs (subvec xs 0 (/ (count xs) 2)) sum ))
                  second_half (future (summingPairs (subvec xs (/ (count xs) 2)) sum ))]
            (concat first_half second_half)       
            )                   
  )]
    (summingPairsHelper xs [])))

File: liuy363/h10/h10.rb
require_relative 'collection'
Pair = Struct.new(:fst,:snd)
def summingPairs(xs, sum)
  the_pairs1 = []
  the_pairs2 = []
  the_pairs = []
  len = xs.length

  reader, writer = IO.pipe 
  
    fork do
     for i in 0..(len-1)
      for j in (i+1)..(len/2)
        if xs[i] + xs[j] <= sum
            writer.puts(Pair.new(xs[i],xs[j]))
        end
      end
    end
  end
  
    fork do
      for i in 0..(len-1)
         for j in (len/2)..(len-1)
             if xs[i] + xs[j] <= sum
                  writer.puts(Pair.new(xs[i],xs[j]))
             end
          end
        end
      end
  Process.waitall
  writer.close
  while message = reader.gets
    the_pairs.push(message)
  end

  return the_pairs
end
